llvm/lib/Transforms/Vectorize/VPlan.cpp

```diff
@@ -1276,6 +1287,19 @@ VPlan *VPlan::duplicate() {
   return NewPlan;
 }
 
+VPBasicBlock *VPlan::getScalarPreheader() {
+  auto *MiddleVPBB =
+      cast<VPBasicBlock>(getVectorLoopRegion()->getSingleSuccessor());
+  if (MiddleVPBB->getNumSuccessors() == 2) {
+    // Order is strict: first is the exit block, second is the scalar preheader.
+    return cast<VPBasicBlock>(MiddleVPBB->getSuccessors()[1]);
+  }
+  if (auto *IRVPBB = dyn_cast<VPBasicBlock>(MiddleVPBB->getSingleSuccessor()))
```

Comment:
dyn_casting to VPBB but calling it `IRVPBB`?
Intention is to return the single successor if it's not an IRVPBB, as that represents the exit block?
```suggestion
  if (isa<VPIRBasicBlock>(MiddleVPBB->getSingleSuccessor())) {
    // Exit block is IRVPBB, scalar preheader is not.
    return nullptr;
  }
  return cast<VPBasicBlock>(MiddleVPBB->getSingleSuccessor());
```